src/core/thread/mesh_forwarder.cpp

```diff
@@ -1617,27 +1617,32 @@ void MeshForwarder::UpdateFramePending()
 
 void MeshForwarder::HandleDataRequest(const Mac::Address &aMacSource, const ThreadMessageInfo &aMessageInfo)
 {
-    Neighbor *neighbor;
-    uint8_t childIndex;
+    Child *child;
 
     // Security Check: only process secure Data Poll frames.
     VerifyOrExit(aMessageInfo.mLinkSecurity, ;);
 
     assert(mMle.GetDeviceState() != Mle::kDeviceStateDetached);
 
-    VerifyOrExit((neighbor = mMle.GetNeighbor(aMacSource)) != NULL, ;);
-    neighbor->mLastHeard = Timer::GetNow();
+    child = mMle.GetChild(aMacSource);
+
+    if (child == NULL)
+    {
+        Neighbor *neighbor;
+
+        VerifyOrExit((neighbor = mMle.GetNeighbor(aMacSource)) != NULL, ;);
```

Comment:
Yes, makes sense. Will change it
I think it will be wrong behavior if we see a data poll from a non-child neighbor anyways, right?